src/mca/gds/ds12/gds_dstore.c

```diff
@@ -3187,8 +3187,8 @@ static pmix_status_t dstore_register_job_info(struct pmix_peer_t *pr,
             return rc;
         }
 
-        PMIX_LIST_FOREACH(rinfo, &ns->ranks, pmix_rank_info_t) {
-            proc.rank = rinfo->pname.rank;
+        for (rank=0; rank < ns->nprocs; rank++) {
+            proc.rank = rank;
```

Comment:
For each nspaces `_store_job_info` will invoke only for ranks started from 0, it looks not correct. Some nspaces do not have rank 0.

Reply:
I can't think of how an nspace would start from a rank other than 0 - can you provide some insight into that scenario?

Reply:
Yeah. I'm not sure I understand that scenario either. If that was the case then the `hash` version should fail as well.

Reply:
@rhc54 when the number of nspaces more than one, each nspace will start rank counting from 0.

I have added printouts of ranks with nspaces:
```patch
diff --git a/src/mca/gds/ds12/gds_dstore.c b/src/mca/gds/ds12/gds_dstore.c
index 580f018..83e9af0 100644
--- a/src/mca/gds/ds12/gds_dstore.c
+++ b/src/mca/gds/ds12/gds_dstore.c
@@ -3187,8 +3187,10 @@ static pmix_status_t dstore_register_job_info(struct pmix_peer_t *pr,
             return rc;
         }
 
+        pmix_output(0, "dstore_register_job_info: nspace=%s, nprocs=%d", ns->nspace, ns->nprocs);
         for (rank=0; rank < ns->nprocs; rank++) {
             proc.rank = rank;
+            pmix_output(0, "_store_job_info for nspace=%s, rank=%d", ns->nspace, proc.rank);
             rc = _store_job_info(&proc);
             if (PMIX_SUCCESS != rc) {
                 PMIX_ERROR_LOG(rc);
```

Here is run `./pmix_test` for two ranks, each rank in different nspace (`smoky_nspace-0` with rank 0, and `smoky_nspace-1` with rank 1):
```
user@fe ~/pmix/master/test $ ./pmix_test -n 2 --ns-dist 1:1 -e ./pmix_init_finalize
[fe:04085] dstore_register_job_info: nspace=smoky_nspace-0, nprocs=1
[fe:04085] _store_job_info for nspace=smoky_nspace-0, rank=0                   <---- rank 0
                                       ^^^^^^^^^^^^^^^^^^^^^
[fe:04085] dstore_register_job_info: nspace=smoky_nspace-1, nprocs=1
[fe:04085] _store_job_info for nspace=smoky_nspace-1, rank=0                   <---- rank 0
                                       ^^^^^^^^^^^^^^^^^^^^^
Client smoky_nspace-0:0 Running...
Client smoky_nspace-1:1 Running...
Client smoky_nspace-1:1 num procs 1
Client smoky_nspace-1:1 PMIx_Get hostname of 0 failed: -17
Client smoky_nspace-0:0 num procs 1
Client smoky_nspace-1:1 Finalizing...
Client smoky_nspace-0:0 hostname of rank 0 is node1
Client smoky_nspace-0:0 Finalizing...
Client smoky_nspace-1:1 PMIx_Finalize successfully completed
Client smoky_nspace-0:0 PMIx_Finalize successfully completed
pmix_test.c:main: Test finished OK!
```
For this case `_store_job_info` was not invoked for rank 1, it stored jdata for ns `smoky_nspace-1` with rank 0, but rank 0 is not assigned for `smoky_nspace-1`.

Reply:
It looks to me like your test is incorrect. Every nspace must start with rank=0 within that nspace. It makes no sense to have an nspace that doesn't have a zero rank.

Reply:
* I agree with @rhc54 that namespace should have a sequential rank numbering starting from 0.
* It looks like PMIx interface currently allows to register clients differently (as this "incorrect" test is working).

@karasevb even given that our implementation is more permissive then it is required I still don't understand why `PMIX_LIST_FOREACH(rinfo, &ns->ranks, pmix_rank_info_t)` doesn't cover all of the ranks. Can you please check why not all of the ranks are in this list.
I'd like to have a root cause.

Reply:
The list only includes the ranks that are _local_ to that PMIx server. We need to have the data from _all_ ranks of the nspace so that things like hostname can be resolved.

Reply:
I take my statements back, I mistakenly believed that we can have multiple namespaces with different ranks. Thanks for the clarification.